lib/sarif_v_2_1_0_util.ml

```diff
@@ -33,40 +34,46 @@ let validate_int64_minimum_one_opt = function
 let validate_int64_minimum_minus_one x =
   if (Int64.compare x (-2L)) > 0 then true else false
 
+let re_guid =
+  Re.Str.regexp "^[0-9a-fA-F]{8}-[0-9a-fA-F]{4}-[1-5][0-9a-fA-F]{3}-[89abAB][0-9a-fA-F]{3}-[0-9a-fA-F]{12}$"
+
 let validate_guid x =
-  let re = Re2.create_exn "^[0-9a-fA-F]{8}-[0-9a-fA-F]{4}-[1-5][0-9a-fA-F]{3}-[89abAB][0-9a-fA-F]{3}-[0-9a-fA-F]{12}$"
-  in Re2.matches re x
+  Re.Str.string_match re_guid x 0
 
-let validate_guid_opt = function
-  | None -> true
-  | Some v -> let re = Re2.create_exn "^[0-9a-fA-F]{8}-[0-9a-fA-F]{4}-[1-5][0-9a-fA-F]{3}-[89abAB][0-9a-fA-F]{3}-[0-9a-fA-F]{12}$"
-              in Re2.matches re v
+let validate_guid_opt o =
+  Option.fold o ~none:true ~some:validate_guid
+
+let re_dotted_quad_file =
+  Re.Str.regexp "^[0-9]+(\\\\.[0-9]+){3}$"
 
 let validate_dotted_quad_file_v x =
-  let re = Re2.create_exn "[0-9]+(\\\\.[0-9]+){3}"
-  in Re2.matches re x
+  Re.Str.string_match re_dotted_quad_file x 0
 
-let validate_dotted_quad_file_v_opt = function
-  | None -> true
-  | Some v -> let re = Re2.create_exn "[0-9]+(\\\\.[0-9]+){3}" in Re2.matches re v
+let validate_dotted_quad_file_v_opt o =
+  Option.fold o ~none:true ~some:validate_dotted_quad_file_v
+
+let re_language =
+  Re.Str.regexp "^[a-zA-Z]{2}(-[a-zA-Z]{2})?$"
 
 let validate_language x =
-  let re = Re2.create_exn "^[a-zA-Z]{2}(-[a-zA-Z]{2})?$"
-  in Re2.matches re x
+  Re.Str.string_match re_language x 0
 
-let validate_language_opt = function
-  | None -> true
-  | Some v -> let re = Re2.create_exn "^[a-zA-Z]{2}(-[a-zA-Z]{2})?$" in Re2.matches re v
+let validate_language_opt x =
+  Option.fold x ~none:true ~some:validate_language
 
 let validate_unique = function
   | [] -> true
-  | lst -> if List.contains_dup lst ~compare:Stdlib.compare then false else true
-
-let validate_unique_opt = function
-  | None -> true
-  | Some v -> match v with
-    | [] -> true
-    | lst -> if List.contains_dup lst ~compare:Stdlib.compare then false else true
+  | cur :: rem ->
+    let rec loop cur rem =
+      not (List.mem cur rem) &&
+      match rem with
+      | [] -> true
+      | cur :: rem -> loop cur rem
+    in
+    loop cur rem
```

Comment:
This is nice, I used core List.contains_dup for a quick and cheap shot which has O(n log n) complexity